docs/architecture/grpc-for-wcf-developers/migrating-duplex-services.md

```diff
@@ -0,0 +1,457 @@
+
+## Bi-directional Streaming
+
+A WCF Full Duplex service allows for asynchronous, real-time messaging in both directions. In the server-streaming example, the client simply initiates a request, then receives a stream of updates. A better version of that service would allow the client to add and remove stocks from the list without needing to stop and create a new subscription. That functionality has been implemented in the [FullStockTickerServer](https://github.com/RendleLabs/grpc-for-wcf-developers/tree/master/FullStockTickerSample/wcf/FullStockTicker) sample.
+
+The `IFullStockTickerService` interface provides three methods:
+
+- `Subscribe` initiates the connection
+- `AddSymbol` adds a stock symbol to watch
+- `RemoveSymbol` removes a symbol
+
+```csharp
+[ServiceContract(SessionMode = SessionMode.Required, CallbackContract = typeof(IFullStockTickerCallback))]
+public interface IFullStockTickerService
+{
+    [OperationContract(IsOneWay = true)]
+    void Subscribe();
+
+    [OperationContract(IsOneWay = true)]
+    void AddSymbol(string symbol);
+
+    [OperationContract(IsOneWay = true)]
+    void RemoveSymbol(string symbol);
+}
+```
+
+The callback interface remains the same.
+
+Implementing this pattern in gRPC is less straightforward, because there are now two streams of data with messages being passed. It is not possible to use multiple methods to implement the Add and Remove operation, but more than one type of message can be passed on a single stream, using either the `Any` type or `oneof` keyword.
+
+For a case where there is a very specific set of types that are acceptable, `oneof` is a better way to go. Use an `ActionMessage` that can hold either an `AddSymbolRequest` or a `RemoveSymbolRequest`.
+
+```protobuf
+message ActionMessage {
+  oneof action {
+    AddSymbolRequest add = 1;
+    RemoveSymbolRequest remove = 2;
+  }
+}
+
+message AddSymbolRequest {
+  string symbol = 1;
+}
+
+message RemoveSymbolRequest {
+  string symbol = 1;
+}
+```
+
+Declare a bi-directional streaming service that takes a stream of `ActionMessage` messages.
+
+```protobuf
+service FullStockTicker {
+  rpc Subscribe (stream ActionMessage) returns (stream StockTickerUpdate);
+}
+```
+
+The implementation for this service is very similar to the previous sample, except the first parameter of the `Subscribe` method is now an `IAsyncStreamReader<ActionMessage>` which can be used to handle the `Add` and `Remove` requests.
+
+```csharp
+public override async Task Subscribe(IAsyncStreamReader<ActionMessage> requestStream, IServerStreamWriter<StockTickerUpdate> responseStream, ServerCallContext context)
```

Comment:
I don't know whether it is applicable here, but I think this is the simplest code for a clean bi-direction streaming service:

https://github.com/grpc/grpc-dotnet/blob/master/examples/Server/Services/RacerService.cs

Reply:
I specifically wanted to use a WPF example for the bi-directional, as I believe that will be applicable to many of the target audience for the book.